api/envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.proto

```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+package envoy.extensions.transport_sockets.proxy_protocol.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/config/core/v3/proxy_protocol.proto";
+
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.transport_sockets.proxy_protocol.v3";
+option java_outer_classname = "UpstreamProxyProtocolProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Upstream Proxy Protocol]
+// [#extension: envoy.transport_sockets.upstream_proxy_protocol]
+
+// Configuration for PROXY protocol socket
+message ProxyProtocolUpstreamTransport {
+  config.core.v3.ProxyProtocolConfig config = 1;
+
+  // The underlying transport socket being wrapped.
+  config.core.v3.TransportSocket transport_socket = 2 [(validate.rules).message = {required: true}];
```

Comment:
I think this is totally fine (I did this for TAP), but just a high-level @envoyproxy/api-shepherds comment here; the fact that we're composing transport sockets via wrapping, rather than pipelining as with listener/HTTP/network filters, seems to be more an accident of design rather than a deliberate decision. Are there any tradeoffs we're making by doing this?

I think the main one I can think of is that we have to implement the transport socket forwarding in each wrapping transport socket, e.g. here, in TAP. It's similar to the has-a vs. is-a tradeoff.

Reply:
> I think this is totally fine (I did this for TAP), but just a high-level @envoyproxy/api-shepherds comment here; the fact that we're composing transport sockets via wrapping, rather than pipelining as with listener/HTTP/network filters, seems to be more an accident of design rather than a deliberate decision. Are there any tradeoffs we're making by doing this?

Initially this is only thought for having SSL/RawBuffer when we first making transport socket pluggable. Though this has some advantages as result (not intentionally):
- wrapping transport socket being able to inspect/modify the wrapped transport socket config, infer more on transport socket initialization (without having FilterState/StreamInfo), this is useful for delegating part of the callback interface to wrapped one.
- wrapping two or more transport socket (was in #7292 which is abandoned, but likely need it for #9577)